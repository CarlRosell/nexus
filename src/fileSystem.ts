import path from "path";

export class FileSystem {
  protected static cachedInstance: FileSystem;

  static getInstance() {
    if (!this.cachedInstance) {
      this.cachedInstance = new FileSystem();
    }
    return this.cachedInstance;
  }

  protected get util() {
    return require("util") as typeof import("util");
  }

  protected get fs() {
    return require("fs") as typeof import("fs");
  }

  protected get mkdir() {
    return this.util.promisify(this.fs.mkdir);
  }

  protected get readFile() {
    return this.util.promisify(this.fs.readFile);
  }

  protected get writeFile() {
    return this.util.promisify(this.fs.writeFile);
  }

  protected get unlink() {
    return this.util.promisify(this.fs.unlink);
  }

  protected ensureAbsolute(filePath: string) {
    if (!path.isAbsolute(filePath)) {
      throw new Error(
        `GraphQL Nexus: Expected an absolute path, saw ${filePath}`
      );
    }
    return filePath;
  }

  getFile(fileName: string) {
    return this.readFile(this.ensureAbsolute(fileName), "utf8");
  }

  /**
   * Unlinks & writes a new file - useful when we are generating the
   * type definitions, as this causes the ts server to know the file has changed.
   * @param path
   * @param toSave
   */
  async replaceFile(filePath: string, toSave: string) {
    if (this.fs.existsSync(this.ensureAbsolute(filePath))) {
      await this.unlink(filePath);
    }
    const dirName = path.dirname(filePath);
    if (!this.fs.existsSync(dirName)) {
      await this.mkdir(dirName, { recursive: true });
    }
    await this.writeFile(filePath, toSave);
  }
}
