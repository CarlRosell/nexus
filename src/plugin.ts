import { SchemaConfig } from "./builder";
import {
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  GraphQLFieldConfig,
} from "graphql";
import { withNexusSymbol, NexusTypes, Omit } from "./definitions/_types";
import { isPromise } from "./utils";
import { NexusObjectTypeConfig } from "./definitions/objectType";
import { NexusOutputFieldConfig } from "./definitions/definitionBlocks";

export class PluginDef {
  constructor(readonly config: PluginConfig) {}
}
withNexusSymbol(PluginDef, NexusTypes.Plugin);

export type PluginVisitAfter = (
  result: any,
  root: any,
  args: any,
  ctx: any,
  info: GraphQLResolveInfo,
  breakVal: typeof BREAK_RESULT_VAL
) => any;

export type PluginVisitBefore = (
  root: any,
  args: any,
  ctx: any,
  info: GraphQLResolveInfo,
  nextVal: typeof NEXT_RESULT_VAL
) => any;

export type PluginVisitor = {
  after?: PluginVisitAfter;
  before?: PluginVisitBefore;
};

export type PluginDefinitionInfo = {
  /**
   * The name of the type we're applying the plugin to
   */
  typeName: string;
  /**
   * The name of the field we're applying the plugin to
   */
  fieldName: string;
  /**
   * The root-level SchemaConfig passed
   */
  nexusSchemaConfig: Omit<SchemaConfig, "types"> & Record<string, unknown>;
  /**
   * The config provided to the Nexus type containing the field.
   * Will not exist if this is a non-Nexus GraphQL type.
   */
  nexusTypeConfig?: Omit<NexusObjectTypeConfig<string>, "definition"> &
    Record<string, unknown>;
  /**
   * The config provided to the Nexus type containing the field.
   * Will not exist if this is a non-Nexus GraphQL type.
   */
  nexusFieldConfig?: Omit<NexusOutputFieldConfig<string, string>, "resolve"> &
    Record<string, unknown>;
  /**
   * Info about the GraphQL Field we're decorating.
   * Always guaranteed to exist, even for non-Nexus GraphQL types
   */
  graphqlFieldConfig: Omit<GraphQLFieldConfig<any, any>, "resolve">;
  /**
   * If we need to collect/reference metadata during the
   * plugin middleware definition stage, you can use this object.
   *
   * An example use would be to collect all fields that have a "validation"
   * property for their input to reference in runtime. After the schema is complete,
   * this object is frozen so it is not abused at runtime.
   */
  mutableObj: Record<string, any>;
};

export interface RootTypingImport {
  /**
   * File path to import the type from.
   */
  path: string;
  /**
   * Name of the type we want to reference in the `path`
   */
  name: string;
  /**
   * Name we want the imported type to be referenced as
   */
  alias?: string;
}

export interface PluginConfig {
  /**
   * A name for the plugin, useful for errors, etc.
   */
  name: string;
  /**
   * A description for the plugin
   */
  description?: string;
  /**
   * Any type definitions we want to add to the schema
   */
  schemaTypes?: string;
  /**
   * Any type definitions we want to add to the type definition option
   */
  typeDefTypes?: string;
  /**
   * Any type definitions we want to add to the field definitions
   */
  fieldDefTypes?: string;
  /**
   * Any extensions to the GraphQLInfoObject (do we need this?)
   */
  // infoExtension?: string;
  /**
   * Any types which should exist as standalone declarations to support this type
   */
  localTypes?: string;
  /**
   * Definition for the plugin. This will be executed against every
   * output type field on the schema.
   */
  pluginDefinition(pluginInfo: PluginDefinitionInfo): PluginVisitor | void;
}

/**
 * A plugin defines configuration which can document additional metadata options
 * for a type definition.
 *
 * Ultimately everything comes down to the "resolve" function
 * which executes the fields, and our plugin system will take that into account.
 *
 * You can specify options which can be defined on the schema,
 * the type or the plugin. The config from each of these will be
 * passed in during schema construction time, and used to augment the field as necessary.
 *
 * You can either return a function, with the new defintion of a resolver implementation,
 * or you can return an "enter" / "leave" pairing which will wrap the pre-execution of the
 * resolver and the "result" of the resolver, respectively.
 */
export function plugin(config: PluginConfig) {
  return new PluginDef(config);
}

/**
 * On "before" plugins, the nextFn allows us to skip to the next resolver
 * in the chain
 */
const NEXT_RESULT_VAL = Object.freeze({});

/**
 * On "after" plugins, the breakerFn allows us to early-return, skipping the
 * rest of the plugin stack. Useful when encountering errors
 */
const BREAK_RESULT_VAL = Object.freeze({});

/**
 * Wraps a resolver with a plugin, "before" the resolver executes.
 * Returns a new resolver, which may subsequently be wrapped
 */
export function wrapPluginsBefore(
  resolver: GraphQLFieldResolver<any, any>,
  beforePlugins: [string, PluginVisitBefore][]
): GraphQLFieldResolver<any, any> {
  return async (root, args, ctx, info) => {
    for (let i = 0; i < beforePlugins.length; i++) {
      const [name, before] = beforePlugins[i];
      let result = before(root, args, ctx, info, NEXT_RESULT_VAL);
      if (isPromise(result)) {
        result = await result;
      }
      if (result === NEXT_RESULT_VAL) {
        continue;
      }
      if (result === undefined) {
        throw new Error(
          `Nexus: Expected return value from plugin ${name}:before, saw undefined`
        );
      }
      return result;
    }
    return resolver(root, args, ctx, info);
  };
}

/**
 * Wraps a resolver with a plugin, "after" the resolver executes.
 * May return a new value for the return type
 */
export function wrapPluginsAfter(
  resolver: GraphQLFieldResolver<any, any>,
  afterPlugins: [string, PluginVisitAfter][]
): GraphQLFieldResolver<any, any> {
  return async (root, args, ctx, info) => {
    let finalResult: any;
    try {
      finalResult = resolver(root, args, ctx, info);
      if (isPromise(finalResult)) {
        finalResult = await finalResult;
      }
    } catch (e) {
      finalResult = e;
    }
    for (let i = 0; i < afterPlugins.length; i++) {
      const [name, after] = afterPlugins[i];
      let returnVal = after(
        finalResult,
        root,
        args,
        ctx,
        info,
        BREAK_RESULT_VAL
      );
      if (isPromise(returnVal)) {
        returnVal = await returnVal;
      }
      if (returnVal === BREAK_RESULT_VAL) {
        return finalResult;
      }
      if (returnVal === undefined) {
        throw new Error(
          `Nexus: Expected return value from plugin ${name}:after, saw undefined`
        );
      }
    }
    return finalResult;
  };
}
