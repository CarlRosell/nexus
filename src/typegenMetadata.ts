import path from "path";
import { Options } from "prettier";
import { GraphQLSchema, lexicographicSortSchema, printSchema } from "graphql";
import { TypegenPrinter } from "./typegen";
import { SDL_HEADER, TYPEGEN_HEADER } from "./lang";
import { typegenAutoConfig } from "./typegenAutoConfig";
import {
  BuilderConfig,
  TypegenInfo,
  NexusSchema,
  SchemaBuilder,
} from "./builder";
import { FileSystem } from "./fileSystem";

export type TypegenFormatFn = (
  content: string,
  type: "types" | "schema"
) => string | Promise<string>;

/**
 * Passed into the SchemaBuilder, this keeps track of any necessary
 * field / type metadata we need to be aware of when building the
 * generated types and/or SDL artifact, including but not limited to:
 */
export class TypegenMetadata {
  protected config: BuilderConfig;
  protected typegenFile: string = "";
  protected sortedSchema: GraphQLSchema;
  protected finalPrettierConfig?: Options;

  constructor(
    protected builder: SchemaBuilder,
    protected nexusSchema: NexusSchema
  ) {
    this.sortedSchema = lexicographicSortSchema(nexusSchema);
    const config = (this.config = builder.getConfig());
    if (config.outputs !== false && config.shouldGenerateArtifacts !== false) {
      if (config.outputs.typegen) {
        this.typegenFile = this.assertAbsolutePath(
          config.outputs.typegen,
          "outputs.typegen"
        );
      }
    }
  }

  getNexusSchema() {
    return this.nexusSchema;
  }

  getBuilder() {
    return this.builder;
  }

  getTypegenFile() {
    return this.typegenFile;
  }

  getSortedSchema() {
    return this.sortedSchema;
  }

  /**
   * Generates the type definitions
   */
  protected async generateTypesFile(): Promise<string> {
    const typegenInfo = await this.getTypegenInfo();
    return new TypegenPrinter(this, typegenInfo).print();
  }

  /**
   * Generates the artifacts of the build based on what we
   * know about the schema and how it was defined.
   */
  async generateArtifacts() {
    if (this.config.outputs) {
      if (this.config.outputs.schema) {
        await this.writeTypeFile(
          "schema",
          this.generateSchemaFile(this.sortedSchema),
          this.config.outputs.schema
        );
      }
      if (this.typegenFile) {
        const value = await this.generateTypesFile();
        await this.writeTypeFile("types", value, this.typegenFile);
      }
    }
  }

  protected async writeTypeFile(
    type: "schema" | "types",
    output: string,
    filePath: string
  ) {
    const [toSave, existing] = await Promise.all([
      this.formatTypegen(output, type),
      FileSystem.getInstance()
        .getFile(filePath)
        .catch(() => ""),
    ]);
    if (toSave !== existing) {
      return FileSystem.getInstance().replaceFile(filePath, toSave);
    }
  }

  /**
   * Generates the schema, adding any directives as necessary
   */
  protected generateSchemaFile(schema: GraphQLSchema): string {
    let printedSchema = printSchema(schema);
    return [SDL_HEADER, printedSchema].join("\n\n");
  }

  async getTypegenInfo(): Promise<TypegenInfo> {
    if (this.config.typegenConfig) {
      if (this.config.typegenAutoConfig) {
        console.warn(
          `Only one of typegenConfig and typegenAutoConfig should be specified, ignoring typegenConfig`
        );
      }
      return this.config.typegenConfig(this.sortedSchema, this.typegenFile);
    }
    if (this.config.typegenAutoConfig) {
      return typegenAutoConfig(this.config.typegenAutoConfig)(
        this.sortedSchema,
        this.typegenFile
      );
    }
    return {
      headers: [TYPEGEN_HEADER],
      imports: [],
      contextType: "any",
      backingTypeMap: {},
    };
  }

  /**
   * Takes "content", which is a string representing the GraphQL schema
   * or the TypeScript types (as indicated by the "fileType"), and formats them
   * with prettier if it is installed.
   * @param content
   * @param fileType
   */
  protected async formatTypegen(content: string, fileType: "types" | "schema") {
    if (this.config.formatTypegen instanceof Function) {
      return this.config.formatTypegen(content, fileType);
    }
    if (!this.config.prettierConfig) {
      return content;
    }
    try {
      const prettierConfig = this.config.prettierConfig;
      const prettier = require("prettier") as typeof import("prettier");
      if (typeof prettierConfig === "string") {
        if (!this.finalPrettierConfig) {
          this.finalPrettierConfig = JSON.parse(
            await FileSystem.getInstance().getFile(prettierConfig)
          ) as Options;
        }
      } else {
        this.finalPrettierConfig = prettierConfig;
      }
      return prettier.format(content, {
        ...this.finalPrettierConfig,
        parser: fileType === "schema" ? "graphql" : "typescript",
      });
    } catch (e) {
      console.error(e);
    }
    return content;
  }

  protected assertAbsolutePath(pathName: string, property: string) {
    if (!path.isAbsolute(pathName)) {
      throw new Error(
        `Expected path for ${property} to be a string, saw ${pathName}`
      );
    }
    return pathName;
  }
}
