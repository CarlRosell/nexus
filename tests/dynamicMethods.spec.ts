import path from "path";
import { makeSchema, objectType, queryType, deferred } from "../src/core";
import { RelayConnectionMethod, CollectionMethod } from "../src/extensions";
import { FileSystem } from "../src/fileSystem";
import { graphql } from "graphql";
import { CatListFixture } from "./_fixtures";

describe("nexus: dynamicOutputMethod", () => {
  const Cat = objectType({
    name: "Cat",
    definition(t) {
      t.id("id");
      t.string("name");
    },
  });

  let spy: jest.SpyInstance;
  beforeEach(() => {
    jest.clearAllMocks();
    spy = jest
      .spyOn(FileSystem.prototype, "replaceFile")
      .mockImplementation(async () => null);
  });

  test("RelayConnectionMethod example", async () => {
    const dfd = deferred();
    const Query = queryType({
      definition(t) {
        // @ts-ignore
        t.relayConnectionField("cats", {
          type: Cat,
          pageInfo: () => ({
            hasNextPage: false,
            hasPreviousPage: false,
          }),
          edges: () =>
            CatListFixture.map((c) => ({ cursor: `Cursor: ${c.id}`, node: c })),
        });
      },
    });
    const schema = makeSchema({
      types: [Query, RelayConnectionMethod],
      outputs: {
        typegen: path.join(__dirname, "test-output.ts"),
        schema: path.join(__dirname, "schema.graphql"),
      },
      shouldGenerateArtifacts: true,
      onReady: dfd.resolve,
    });
    await dfd.promise;
    expect(spy.mock.calls[0]).toMatchSnapshot();
    expect(spy.mock.calls[1]).toMatchSnapshot();
    expect(
      await graphql(
        schema,
        `
          {
            cats {
              edges {
                node {
                  id
                  name
                }
              }
              pageInfo {
                hasNextPage
                hasPreviousPage
              }
            }
          }
        `
      )
    ).toMatchSnapshot();
  });

  test("CollectionMethod example", async () => {
    const dynamicOutputMethod = queryType({
      definition(t) {
        // @ts-ignore
        t.collectionField("cats", {
          type: Cat,
          totalCount: () => CatListFixture.length,
          nodes: () => CatListFixture,
        });
      },
    });

    const dfd = deferred();

    const schema = makeSchema({
      types: [dynamicOutputMethod, CollectionMethod],
      outputs: {
        typegen: path.join(__dirname, "test-output"),
        schema: path.join(__dirname, "schema.graphql"),
      },
      onReady: dfd.resolve,
      shouldGenerateArtifacts: true,
    });

    await dfd.promise;
    expect(spy.mock.calls[0]).toMatchSnapshot();
    expect(spy.mock.calls[1]).toMatchSnapshot();

    expect(
      await graphql(
        schema,
        `
          {
            cats {
              totalCount
              nodes {
                id
                name
              }
            }
          }
        `
      )
    ).toMatchSnapshot();
  });

  test("CollectionMethod example with string type ref", () => {
    makeSchema({
      types: [
        queryType({
          definition(t) {
            // @ts-ignore
            t.collectionField("cats", {
              type: "Cat",
              totalCount: () => CatListFixture.length,
              nodes: () => CatListFixture,
            });
          },
        }),
        CollectionMethod,
      ],
      outputs: false,
    });
  });

  test("RelayConnectionMethod example with string type ref", async () => {
    makeSchema({
      types: [
        queryType({
          definition(t) {
            // @ts-ignore
            t.relayConnectionField("cats", {
              type: "Cat",
              pageInfo: () => ({
                hasNextPage: false,
                hasPreviousPage: false,
              }),
              edges: () =>
                CatListFixture.map((c) => ({
                  cursor: `Cursor: ${c.id}`,
                  node: c,
                })),
            });
          },
        }),
        RelayConnectionMethod,
      ],
      outputs: false,
    });
  });
});
