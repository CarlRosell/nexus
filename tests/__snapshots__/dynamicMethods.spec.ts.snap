// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nexus: dynamicOutputMethod CollectionMethod example 1`] = `
Array [
  "/Users/tgriesser/Github/oss/nexus/nexus/tests/schema.graphql",
  "### This file was autogenerated by GraphQL Nexus
### Do not make changes to this file directly


type Cat {
  id: ID!
  name: String!
}

type CatCollection {
  nodes: [Cat!]!
  totalCount: Int!
}

type Query {
  cats(page: Int, perPage: Int): CatCollection!
}
",
]
`;

exports[`nexus: dynamicOutputMethod CollectionMethod example 2`] = `
Array [
  "/Users/tgriesser/Github/oss/nexus/nexus/tests/test-output",
  "/**
 * This file was automatically generated by GraphQL Nexus
 * Do not make changes to this file directly
 */


import { core } from \\"nexus\\"

declare global {
  interface NexusGenCustomOutputMethods<TypeName extends string> {
    collectionField<FieldName extends string>(fieldName: FieldName, opts: {
      type: NexusGenObjectNames | NexusGenInterfaceNames | core.NexusObjectTypeDef<string> | core.NexusInterfaceTypeDef<string>,
      nodes: core.SubFieldResolver<TypeName, FieldName, \\"nodes\\">,
      totalCount: core.SubFieldResolver<TypeName, FieldName, \\"totalCount\\">,
      args?: core.ArgsRecord,
      nullable?: boolean,
      description?: string
    }): void;
  }
}

declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenRootTypes {
  Cat: { // root type
    id: string; // ID!
    name: string; // String!
  }
  CatCollection: { // root type
    nodes: NexusGenRootTypes['Cat'][]; // [Cat!]!
    totalCount: number; // Int!
  }
  Query: {};
  String: string;
  Int: number;
  Float: number;
  Boolean: boolean;
  ID: string;
}

export interface NexusGenAllTypes extends NexusGenRootTypes {
}

export interface NexusGenFieldTypes {
  Cat: { // field return type
    id: string; // ID!
    name: string; // String!
  }
  CatCollection: { // field return type
    nodes: NexusGenRootTypes['Cat'][]; // [Cat!]!
    totalCount: number; // Int!
  }
  Query: { // field return type
    cats: NexusGenRootTypes['CatCollection']; // CatCollection!
  }
}

export interface NexusGenArgTypes {
  Query: {
    cats: { // args
      page?: number | null; // Int
      perPage?: number | null; // Int
    }
  }
}

export interface NexusGenAbstractResolveReturnTypes {
}

export interface NexusGenInheritedFields {}

export type NexusGenObjectNames = \\"Cat\\" | \\"CatCollection\\" | \\"Query\\";

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = \\"Boolean\\" | \\"Float\\" | \\"ID\\" | \\"Int\\" | \\"String\\";

export type NexusGenUnionNames = never;

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  allTypes: NexusGenAllTypes;
  inheritedFields: NexusGenInheritedFields;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractResolveReturn: NexusGenAbstractResolveReturnTypes;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginSchemaConfig {
  }
}",
]
`;

exports[`nexus: dynamicOutputMethod CollectionMethod example 3`] = `
Object {
  "data": Object {
    "cats": Object {
      "nodes": Array [
        Object {
          "id": "Cat:1",
          "name": "Felix",
        },
        Object {
          "id": "Cat:2",
          "name": "Booker",
        },
      ],
      "totalCount": 2,
    },
  },
}
`;

exports[`nexus: dynamicOutputMethod RelayConnectionMethod example 1`] = `
Array [
  "/Users/tgriesser/Github/oss/nexus/nexus/tests/schema.graphql",
  "### This file was autogenerated by GraphQL Nexus
### Do not make changes to this file directly


type Cat {
  id: ID!
  name: String!
}

type CatEdge {
  cursor: ID!
  node: Cat!
}

type CatRelayConnection {
  edges: [CatEdge!]!
  pageInfo: ConnectionPageInfo!
}

type ConnectionPageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
}

type Query {
  cats(after: String, before: String, first: Int, last: Int): CatRelayConnection!
}
",
]
`;

exports[`nexus: dynamicOutputMethod RelayConnectionMethod example 2`] = `
Array [
  "/Users/tgriesser/Github/oss/nexus/nexus/tests/test-output.ts",
  "/**
 * This file was automatically generated by GraphQL Nexus
 * Do not make changes to this file directly
 */


import { core } from \\"nexus\\"

declare global {
  interface NexusGenCustomOutputMethods<TypeName extends string> {
    relayConnectionField
    <FieldName extends string>(fieldName: FieldName, opts: {
      type: NexusGenObjectNames | NexusGenInterfaceNames | core.NexusObjectTypeDef<string> | core.NexusInterfaceTypeDef<string>,
      edges: core.SubFieldResolver<TypeName, FieldName, \\"edges\\">,
      pageInfo: core.SubFieldResolver<TypeName, FieldName, \\"pageInfo\\">,
      args?: Record<string, core.NexusArgDef<string>>,
      nullable?: boolean,
      description?: string
    }): void
  
  }
}

declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenRootTypes {
  Cat: { // root type
    id: string; // ID!
    name: string; // String!
  }
  CatEdge: { // root type
    cursor: string; // ID!
    node: NexusGenRootTypes['Cat']; // Cat!
  }
  CatRelayConnection: { // root type
    edges: NexusGenRootTypes['CatEdge'][]; // [CatEdge!]!
    pageInfo: NexusGenRootTypes['ConnectionPageInfo']; // ConnectionPageInfo!
  }
  ConnectionPageInfo: { // root type
    hasNextPage: boolean; // Boolean!
    hasPreviousPage: boolean; // Boolean!
  }
  Query: {};
  String: string;
  Int: number;
  Float: number;
  Boolean: boolean;
  ID: string;
}

export interface NexusGenAllTypes extends NexusGenRootTypes {
}

export interface NexusGenFieldTypes {
  Cat: { // field return type
    id: string; // ID!
    name: string; // String!
  }
  CatEdge: { // field return type
    cursor: string; // ID!
    node: NexusGenRootTypes['Cat']; // Cat!
  }
  CatRelayConnection: { // field return type
    edges: NexusGenRootTypes['CatEdge'][]; // [CatEdge!]!
    pageInfo: NexusGenRootTypes['ConnectionPageInfo']; // ConnectionPageInfo!
  }
  ConnectionPageInfo: { // field return type
    hasNextPage: boolean; // Boolean!
    hasPreviousPage: boolean; // Boolean!
  }
  Query: { // field return type
    cats: NexusGenRootTypes['CatRelayConnection']; // CatRelayConnection!
  }
}

export interface NexusGenArgTypes {
  Query: {
    cats: { // args
      after?: string | null; // String
      before?: string | null; // String
      first?: number | null; // Int
      last?: number | null; // Int
    }
  }
}

export interface NexusGenAbstractResolveReturnTypes {
}

export interface NexusGenInheritedFields {}

export type NexusGenObjectNames = \\"Cat\\" | \\"CatEdge\\" | \\"CatRelayConnection\\" | \\"ConnectionPageInfo\\" | \\"Query\\";

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = \\"Boolean\\" | \\"Float\\" | \\"ID\\" | \\"Int\\" | \\"String\\";

export type NexusGenUnionNames = never;

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  allTypes: NexusGenAllTypes;
  inheritedFields: NexusGenInheritedFields;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractResolveReturn: NexusGenAbstractResolveReturnTypes;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginSchemaConfig {
  }
}",
]
`;

exports[`nexus: dynamicOutputMethod RelayConnectionMethod example 3`] = `
Object {
  "data": Object {
    "cats": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "Cat:1",
            "name": "Felix",
          },
        },
        Object {
          "node": Object {
            "id": "Cat:2",
            "name": "Booker",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
    },
  },
}
`;
